packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts

```diff
@@ -68,226 +68,133 @@ export const getDaysInWeekStr = <TDate>(utils: MuiPickersAdapter<TDate>, format:
 };
 
-export const adjustDateSectionValue = <TDate>(
+export const cleanTrailingZeroInNumericSectionValue = <TDate>(
   utils: MuiPickersAdapter<TDate>,
-  date: TDate,
-  dateSectionName: MuiDateSectionName,
-  keyCode: AvailableAdjustKeyCode,
-): TDate => {
-  const delta = getDeltaFromKeyCode(keyCode);
-  const isStart = keyCode === 'Home';
-  const isEnd = keyCode === 'End';
+  format: string,
+  value: string,
+) => {
+  const size = utils.formatByString(utils.date()!, format).length;
+  let cleanValue = value;
+
+  // We remove the trailing zeros
+  cleanValue = Number(cleanValue).toString();
 
+  // We add enough trailing zeros to fill the section
+  while (cleanValue.length < size) {
+    cleanValue = `0${cleanValue}`;
+  }
+
+  return cleanValue;
+};
+
+export const getLetterEditingOptions = <TDate>(
+  utils: MuiPickersAdapter<TDate>,
+  dateSectionName: MuiDateSectionName,
+  format: string,
+) => {
   switch (dateSectionName) {
-    case 'day': {
-      if (isStart) {
-        return utils.startOfMonth(date);
-      }
-      if (isEnd) {
-        return utils.endOfMonth(date);
-      }
-      return utils.addDays(date, delta);
-    }
-    case 'weekDay': {
-      if (isStart) {
-        return utils.startOfWeek(date);
-      }
-      if (isEnd) {
-        return utils.endOfWeek(date);
-      }
-      return utils.addDays(date, delta);
-    }
     case 'month': {
-      if (isStart) {
-        return utils.startOfYear(date);
-      }
-      if (isEnd) {
-        return utils.endOfYear(date);
-      }
-      return utils.addMonths(date, delta);
+      return utils
+        .getMonthArray(utils.date()!)
+        .map((month) => utils.formatByString(month, format!));
     }
-    case 'year': {
-      return utils.addYears(date, delta);
+
+    case 'weekDay': {
+      return getDaysInWeekStr(utils, format);
     }
+
     case 'meridiem': {
-      return utils.addHours(date, (delta > 0 ? 1 : -1) * 12);
-    }
-    case 'hours': {
-      if (isStart) {
-        return utils.startOfDay(date);
-      }
-      if (isEnd) {
-        return utils.endOfDay(date);
-      }
-      return utils.addHours(date, delta);
-    }
-    case 'minutes': {
-      if (isStart) {
-        return utils.setMinutes(date, 0);
-      }
-      if (isEnd) {
-        return utils.setMinutes(date, 59);
-      }
-      return utils.addMinutes(date, delta);
-    }
-    case 'seconds': {
-      if (isStart) {
-        return utils.setSeconds(date, 0);
-      }
-      if (isEnd) {
-        return utils.setSeconds(date, 59);
-      }
-      return utils.addSeconds(date, delta);
+      const now = utils.date()!;
+      return [utils.endOfDay(now), utils.startOfDay(now)].map((date) =>
+        utils.formatByString(date, format),
+      );
     }
+
     default: {
-      return date;
+      return [];
     }
   }
 };
 
-export const adjustInvalidDateSectionValue = <TDate, TSection extends FieldSection>(
+export const adjustSectionValue = <TDate, TSection extends FieldSection>(
```

Comment:
This function changes a lot.
It now uses a similar approach to the character editing.
We never try to parse a date (except to get today's year).

Instead for digit editing we rely on the boundaries and for letter editing on the option list.

Reply:
The old approach had problems with days of the month for example, because by parsing today's date, we had a "max" of "28" when using the field in February.

The new version should be more robust since it uses the boundaries which have a smarter behavior.